### PR TITLE
Implement transaction balancing (prototype)

### PR DESCRIPTION
**Pull request by ajuggler**
_Objective:_  implement transaction balancing algorithm.

### CODE REVIEW

**Comment by ajuggler** [comment]
The purpose of last commit is to clarify/exemplify the requirements of task [Implement transaction balancing algorithm](https://github.com/zkFold/zkfold-cardano/issues/6#issue-2268695831).  Also, we want clarify whether cardano-cli's command `calculate-min-required-utxo` already implements what we need.

With modifications in last commit, *automatic balancing* depending on inputs is already achieved in the context of the plonk verifier example.  The inputs are as follows:

- In '01-init-transaction.sh', relevant inputs are serialized scripts `plonkVerifier.plutus` and `forwardingMint.plutus`.
- In '02-transfer-transaction.sh', relevant input is datum `datumPlonk.cbor`.
- In '03-minting-transaction.sh', relevant input are `policyid` + `tokenname` for plonkVerifier and verification input, respectively.  

I guess that, for tokenname, we would want to use a generic example so as to avoid computation of *input*.

File: e2e-test/plonk/01-init-transaction.sh [hunk 1]
```diff
@@ -35,14 +35,34 @@ cardano-cli conway address build \
     --out-file "$keypath/plonkVerifier.addr" \
     --testnet-magic $magic
 
+#-------------------------------- :min-required-utxo: --------------------------
+
+cardano-cli conway query protocol-parameters \
+  --testnet-magic $magic \
+  --out-file $assets/protocol.json
+
+plonkVerifierSize=$(stat -c%s "$assets/plonkVerifier.plutus")
```
Comment (ajuggler) [inline e2e-test/plonk/01-init-transaction.sh:44]: Actually not needed.  Also, line 46 is not needed.

**Comment by ajuggler** [comment]
Shell script `compute-min-fee` estimates the `minFee` for executing `symbolicVerifer` on a given set of data inputs (`txInfoInputs`, `txInfoReferenceInputs`, `txInfoOutputs` and `txInfoValidRange`) recorded in file `inputs.json`.  Estimation is based on linear model constructed with aid of plutus scripts `pubInput` and `symbolicVerifierBench1` (file PubInput.hs).

**Comment by ajuggler** [comment]
Relevant directory is `/benchs/bench-emptycircuit`, which contains a README summarizing the test so far.

**Comment by ajuggler** [comment]
Note that [`tautologyVerificationBytes`](https://github.com/zkFold/zkfold-cardano/blob/94fe56e5d8ed97300dc2258bf5924ac5bee8b8ab/src/ZkFold/Cardano/Benchs/EmptyCircuit.hs#L19) was defined with "emptyCircuit":
```haskell
mempty {acOutput = V.empty} :: ArithmeticCircuit Fr (Vector 1) (Vector 0)
```

**Comment by ajuggler** [comment]
To test [`tautologyVerificationBytes`](https://github.com/zkFold/zkfold-cardano/blob/752a82fbf92c702a1872cba2d534f5d053555703/src/ZkFold/Cardano/Benchs/EmptyCircuit.hs#L19), simply execute
```shell
cabal run bench-emptycircuit
```

(Active directory is irrelevant.)

**Comment by ajuggler** [comment]
Follow instructions in [README](https://github.com/zkFold/zkfold-cardano/blob/e2e-balancing-tx/e2e-test/README.md#symbolic-balancing).

Terminal output:
```shell
antonio:e2e-test$ ./symbolic-balancing/02-balancing-parse-input.sh 
Input UTxO from Alice:

Right [TxInInfo {txInInfoOutRef = TxOutRef {txOutRefId = 28f5dd8345a0ca8343af949babe74a62fea9634b8b7ee0ed178055ea903e32c8, txOutRefIdx = 1}, txInInfoResolved = TxOut {txOutAddress = Address {addressCredential = PubKeyCredential d6afdb7c4ac7c5a211cfbed049011b7f0a4badacbdd2b8aa04359d5c, addressStakingCredential = Nothing}, txOutValue = Value {getValue = Map {unMap = [(,Map {unMap = [("",950066712)]})]}}, txOutDatum = NoOutputDatum, txOutReferenceScript = Nothing}}]

Input UTxO from SymbolicVerifier:

Right [TxInInfo {txInInfoOutRef = TxOutRef {txOutRefId = 28f5dd8345a0ca8343af949babe74a62fea9634b8b7ee0ed178055ea903e32c8, txOutRefIdx = 0}, txInInfoResolved = TxOut {txOutAddress = Address {addressCredential = ScriptCredential eff4bde20f815128049ffecadbff04c58920e0cabb94662c9e277fa2, addressStakingCredential = Nothing}, txOutValue = Value {getValue = Map {unMap = [(,Map {unMap = [("",10000000)]})]}}, txOutDatum = OutputDatum (Datum {getDatum = Constr 0 []}), txOutReferenceScript = Nothing}}]

Data:

Constr 0 [Constr 0 [B "(\245\221\131E\160\202\131C\175\148\155\171\231Jb\254\169cK\139~\224\237\ETB\128U\234\144>2\200",I 1],Constr 0 [Constr 0 [Constr 0 [B "\214\175\219|J\199\197\162\DC1\207\190\208I\SOH\ESC\DEL\nK\173\172\189\210\184\170\EOT5\157\\"],Constr 1 []],Map [(B "",Map [(B "",I 950066712)])],Constr 0 [],Constr 1 []]]

Serialised data:

"\216y\159\216y\159X (\245\221\131E\160\202\131C\175\148\155\171\231Jb\254\169cK\139~\224\237\ETB\128U\234\144>2\200\SOH\255\216y\159\216y\159\216y\159X\FS\214\175\219|J\199\197\162\DC1\207\190\208I\SOH\ESC\DEL\nK\173\172\189\210\184\170\EOT5\157\\\255\216z\128\255\161@\161@\SUB8\160\222\CAN\216y\128\216z\128\255\255"

Verifier's input: F 14863051969094712002847497831427595412153433551950329341365388888186

Generating proof...

Retrieving funds from Symbolic [exec units]...

Estimated transaction fee: Coin 653729
```

**Comment by ajuggler** [comment]
Balancing prototype now works as expected.  To test code, follow instructions in [README](https://github.com/zkFold/zkfold-cardano/blob/e2e-balancing-tx/e2e-test/README.md#symbolic-balancing).

File: backends/symbolic-balancing/balancing-init-transaction/Main.hs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+  createDirectoryIfMissing True $ path </> "test-data"
+  createDirectoryIfMissing True $ path </> "assets"
+
+  x           <- generate arbitrary
+  ps          <- generate arbitrary
+
+  let contract = IdentityCircuitContract x ps
+
+  BL.writeFile (path </> "test-data" </> "plonk-raw-contract-data.json") $ encode contract
```
Comment (zlonast) [inline backends/symbolic-balancing/balancing-init-transaction/Main.hs:53]: i prefer `symbolic-contract-data.json`

File: backends/symbolic-balancing/balancing-parse-input/Backend/JsonToData.hs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+import           Data.Text            (Text)
+import qualified Data.Text            as T
+import qualified Data.Text.Encoding   as TE
+import           PlutusLedgerApi.V3   as V3
+import           PlutusTx             (CompiledCode)
+import           Prelude              (String, const, maybe, return, show, ($), (++), (.))
+import           Text.Read            (readEither)
+
+-- Function to parse JSON and convert to a list of TxInInfo
+parseJsonToTxInInfoList :: [Maybe (CompiledCode a)] -> BL.ByteString -> Either String [TxInInfo]
```
Comment (zlonast) [inline backends/symbolic-balancing/balancing-parse-input/Backend/JsonToData.hs:26]: I think a similar parser should be in the Atlas 🤔

File: backends/symbolic/symbolic-init-transaction/Main01.bak [hunk 4]
```diff
@@ -0,0 +1,204 @@
+module Main where
```
Comment (zlonast) [inline backends/symbolic/symbolic-init-transaction/Main01.bak:1]: Does this seem unnecessary?

File: backends/symbolic/symbolic-withdraw-transaction/#Main.hs# [hunk 5]
```diff
@@ -0,0 +1,33 @@
+module Main where
```
Comment (zlonast) [inline backends/symbolic/symbolic-withdraw-transaction/#Main.hs#:1]: Does this seem unnecessary?

File: backends/symbolic-balancing/balancing-init-transaction/Main.hs (hunk 2 above)
Comment (vlasin) [reply backends/symbolic-balancing/balancing-init-transaction/Main.hs:53]: This should be refactored anyway. As I said before, x will not be accessible to smart contract developers, only the Plonk/Plonkup setup objects!

Comment (ajuggler) [reply backends/symbolic-balancing/balancing-init-transaction/Main.hs:53]: Done.

File: backends/symbolic-balancing/balancing-parse-input/Backend/JsonToData.hs (hunk 3 above)
Comment (ajuggler) [reply backends/symbolic-balancing/balancing-parse-input/Backend/JsonToData.hs:26]: Yes, with Atlas all of this should be simpler.

File: backends/symbolic/symbolic-init-transaction/Main01.bak (hunk 4 above)
Comment (ajuggler) [reply backends/symbolic/symbolic-init-transaction/Main01.bak:1]: Deleted this old backup file.

File: backends/symbolic/symbolic-withdraw-transaction/#Main.hs# (hunk 5 above)
Comment (ajuggler) [reply backends/symbolic/symbolic-withdraw-transaction/#Main.hs#:1]: Deleted.

File: backends/symbolic-balancing/balancing-parse-input/Main.hs [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+main :: IO ()
+main = do
+  currentDir <- getCurrentDirectory
+  let path = case takeFileName currentDir of
+        "symbolic-balancing" -> ".." </> ".."
+        "backends"           -> ".."
+        "e2e-test"           -> ".."
+        _                    -> "."
+
+  IdentityCircuitContract x ps <- fromJust . decode <$> BL.readFile (path </> "test-data" </> "plonk-raw-contract-data.json")
```
[the remark this answers is not in the document]
Comment (ajuggler) [reply backends/symbolic-balancing/balancing-parse-input/Main.hs:41]: Done.

File: benchs/bench-pubinput/scripts/keys/alice.addr [hunk 7]
```diff
@@ -0,0 +1 @@
+addr_test1vz36jgww40x7cufqya5tt7ycvedm5curkjjguzz7x09edmsnqwgwf
```
[the remark this answers is not in the document]
Comment (ajuggler) [reply benchs/bench-pubinput/scripts/keys/alice.addr:1]: Removed directory 'benchs/bench-pubinput', as it only contained outdated material that had been refactored elsewhere.

**Review by zlonast — APPROVED** [review]

**Pull request merged**